Variant tables

JureZmrzlikar opened the pull request.

AnzeLovse commented on src/resdk/tables/variant.py:
```diff
@@ -0,0 +1,157 @@
+
+    @property
+    @lru_cache()
+    def mutation_subset_genes(self) -> pd.DataFrame:
+        """Return genes for which variants are reported."""
+        return self._data[0].output["genes"]
```
The docstring could be extended because it could be misleading for someone not familiar with `mutations-table`. For example, we use gene `FOO` and its mutation `Val10` as an input to the `mutations-table`. After using the function and getting `["FOO"]` as a list of genes one could naively assume that all other mutations of this gene were considered as well.

AnzeLovse commented on src/resdk/tables/variant.py:
```diff
@@ -0,0 +1,157 @@
+        # If all samples are missing a variant, remove it
+        df = df.dropna(axis=1, how="all")
+        # No mutation detected means WT, ergo encode 0
+        return df.fillna(0)
```
What happens if we use different genes for different samples?

If we had `mutation-tables` for one gene on one sample and another gene one the other sample wouldn't `df.fillna(0)` transform non-analysed genes to a WT.

AnzeLovse reviewed (commented):
Overall this looks great! The tests pass and I successfully get the results from the toy collection.

OtonicarJan commented on src/resdk/tables/variant.py:
```diff
@@ -0,0 +1,157 @@
+        Construct index of the variants table.
+
+        Index should have the form:
+        <chr>_<position>_<snp_change>_<aminoacid_pos_change>
```
It would be great if the gene name is also added to index if we run the analysis for multiple genes at once.

OtonicarJan commented on src/resdk/tables/variant.py:
```diff
@@ -0,0 +1,157 @@
+        elif data_type == self.DEPTH:
+            # Depth, as computed by GATK is reported by "DP" column
+            s = sample_data["DP"]
+        elif data_type == self.COVERAGE:
```
When I ran the test in the toy collection, I get this output: 
```
>>> vt.coverage
index      chr12_25245350_C>T_Gly12Asp
sample_id                             
9073                               0.0
9075                               91.0
```
The coverage for the sample 9073 should be reported (96). Coverage ("Total_depth") is always in the output in mutations table, so it should also be in the output here even if there is no mutation present.

OtonicarJan commented on src/resdk/tables/variant.py:
```diff
@@ -0,0 +1,157 @@
+    process_type = "data:mutationstable"
+    VARIANTS = "variants"
+    DEPTH = "depth"
+    COVERAGE = "coverage"
```
Reporting coverage is OK, but reporting base frequencies would offer even more insight about the variant. If we indeed report base frequencies then maybe coverage report can be removed.

OtonicarJan reviewed (commented):
Looks good. Left few comments.

JureZmrzlikar commented on src/resdk/tables/variant.py:
```diff
@@ -0,0 +1,157 @@
+        sample_data.set_index("index", inplace=True)
+
+        if data_type == self.VARIANTS:
+            s = sample_data.apply(self.encode_mutation, axis=1, result_type="reduce")
```
This happens because there are `NaN` values present for this particluar variant. NaN's are floats, so all values in a given column are casted to float. I would rather keep the possiblity of having `np.nan`'s than to cast this to integers and present missing values with pandas.NA (which can represent missing integers).

AnzeLovse replied:
I agree, keeping in the `NaN` values makes sense and is more informative.

JureZmrzlikar commented on src/resdk/tables/variant.py:
```diff
@@ -0,0 +1,157 @@
+            ).group(1, 2)
+        except (AttributeError, TypeError):
+            # TypeError is raised when row["SAMPLENAME1.GT"] is np.nan / None
+            # AttributeError is raised when there is no match
```
I don't really know. I wanted to handle the case of empty string (`""`), but we might want to reconsider this.

AnzeLovse replied:
It is nice to handle that as well. Perhaps returning `np.nan` would be more appropriate in the case of an empty string.

JureZmrzlikar commented:
Hi @AnzeLovse and @OtonicarJan , I'm asking for (at least) another review pass.

After trying to apply your fixes i saw the need to optimize / generalize things a bit more... So here you have. IMO I've addressed your comments but the amount of changes looks more like a rewrite. Feel free to ask questions. Or just point our parts that are unclear - we probably want to document those in more depth. Maybe a good place to start reviewing is the docs part to get the general "why" behind some of the new functionality.

AnzeLovse reviewed (approved):
Amazing effort! The expended documentation is great and the code seems more robust now. Remember to merge WIP commits.

OtonicarJan commented on src/resdk/tables/variant.py:
```diff
@@ -0,0 +1,424 @@
+            fields=self.DATA_FIELDS,
+        ):
+            # 1 Filter by mutations, if required
+            if self.mutations:
```
Currently, if you call `VariantTables` and set a certain "gene: aminoacid" combination that is not present in the target samples as an input to `mutations`, you will get an error `Collection <collection> has no data:mutationstable data!`, even though data objects with type `data:mutationstable` exist in this collection. It would be better if this error messages were more informative. Example of an error: 

```
vt = VariantTables(c, mutations=["TP53: Ser"])
vt.variants
ValueError: Collection <collection> has no data:mutationstable data!
```

OtonicarJan commented on src/resdk/tables/variant.py:
```diff
@@ -0,0 +1,424 @@
+        to_discard = list()
+
+        for _, group in groupby(df.columns, key=chrom_pos):
+            # Sorting moves fake variant to the first position
```
There is an issue if you first call `VariantTables` with `discard_fakes` set to default (False) and then try to call `VariantTables` again but with `discard_fakes` set to True. "fakes" will not be removed in the second call, you first have to clear the cache. The default value for `discard_fakes` could be set to True.

OtonicarJan reviewed (approved):
Amazing effort! Left two suggestions for improvement.
